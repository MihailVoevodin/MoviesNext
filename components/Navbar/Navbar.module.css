.navbarList {
    display: flex;
    flex-direction: column;
}

.navbarList a {
    text-align: center;
    padding: 10px;
    margin: 10px;
}