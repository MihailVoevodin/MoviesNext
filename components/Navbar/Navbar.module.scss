@import 'styles/variables';

.navbarList {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    a {
        transition: all .5s;
        border-radius: 15px;
        text-align: center;
        padding: 10px;
        margin: 10px;
        &:hover {
            background-color: $secondary;
        }
    }
}

.searchContainer {
    position: relative;
}

.searchInput {
    width: 0;
    opacity: 0;
    animation: anim 1s linear forwards;
    margin-right: 10px;
    border-color: $secondary;
    &:focus {
        border-color: $secondary;
    }
    &::placeholder {
        opacity: 0.3;
        color: $secondary;
    }
}

.searchBtn {
    color: $secondary;
}

.searchMovies {
    color: $primary;
    position: absolute;
    width: 500px;
    height: 500px;
    top: 35px;
    border-radius: 6px;
    overflow-y: auto;
    background-color: $white;
    z-index: 2;
    padding: 10px;
    a {
        border-radius: 0;
        padding: 0;
        margin: 10px;
    }
}

.searchMovie {
    display: flex;
    align-items: center;
    transition: all 1s;
    text-align: start;
    img {
        margin-right: 30px;
    }
    &:hover {
        background-color: $lightGrey;
    }
}

.searchMovieText {
    display: flex;
    opacity: .5;
    .searchMovieRating {
        margin-right: 20px;
    }
}

@keyframes anim{
    50%{
        opacity: 0.5;
        width: 250px;
    }
    100%{
        opacity: 1;
        width: 500px;
    }
}
