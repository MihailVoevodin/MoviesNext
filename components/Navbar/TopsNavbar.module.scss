@import 'styles/variables';

.topsNavbarContainer {
    margin: 0 auto;
    display: flex;
    background-color: $darkGrey;
    justify-content: center;
    border-top: 2px solid #000000;
    a {
        display: inline-block;
        margin: 20px 0;
        padding: 5px 10px;
        border-radius: 10px;
        transition: all .5s;
        &:not(:first-child) {
            margin-left: 20px;
        }
        &:hover {
            background-color: $secondary;
        }
    }
}

.active {
    background-color: $secondary;
}
