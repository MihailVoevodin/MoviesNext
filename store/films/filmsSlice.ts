import {createAsyncThunk, createSlice, PayloadAction} from '@reduxjs/toolkit';
import {IMovie} from 'Common/Models';
import {TFilmRating} from 'Common/Models';
import {Services} from 'Common/Services';

/**
 * @param top250PageId - Номер страницы топ 250 фильмов.
 * @param top100PageId - Номер страницы топ 100 популярных фильмов.
 * @param topAwaitPageId - Номер страницы топа ожидаемых фильмов.
 * @param imagesPageId - Номер страницы изображений.
 * @param reviewsPageId - Номер страницы рецензий.
 */
export interface IFilmsState {
    searchMovies: IMovie[];
    sequelsAndPrequels: IMovie[];
    activeTabName: string;
    pagesId: {
        top250: number;
        top100: number;
        topAwait: number;
        images: number;
        reviews: number;
    };
    filmRating: TFilmRating;
    isLoading: boolean;
    isError: boolean;
}

const initialState: IFilmsState = {
    searchMovies: [],
    sequelsAndPrequels: [],
    activeTabName: 'top250movies',
    pagesId: {
        top250: 1,
        top100: 1,
        topAwait: 1,
        images: 1,
        reviews: 1,
    },
    filmRating: {},
    isLoading: false,
    isError: false,
};

export const loadMoviesBySearch = createAsyncThunk('films/loadMoviesBySearch', async (keyword: string, {rejectWithValue}) => {
    const response = await Services.getMoviesBySearch(keyword);

    if (response.status !== 200) {
        return rejectWithValue('Server Error!');
    }

    return response.data.films;
});

export const loadMovieSequelsAndPrequels = createAsyncThunk(
    'films/loadMovieSequelsAndPrequels',
    async (movieId: number, {rejectWithValue}) => {
        const response = await Services.getMovieSequelsAndPrequels(movieId);

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

/**
 * Срез списка фильмов.
 */
const filmsSlice = createSlice({
    name: 'films',
    initialState: initialState,
    reducers: {
        setActiveTabName(state, action: PayloadAction<string>) {
            state.activeTabName = action.payload;
        },
        setTop250PageId(state, action: PayloadAction<number>) {
            state.pagesId.top250 = action.payload;
        },
        setTop100PageId(state, action: PayloadAction<number>) {
            state.pagesId.top100 = action.payload;
        },
        setTopAwaitPageId(state, action: PayloadAction<number>) {
            state.pagesId.topAwait = action.payload;
        },
        setImagesPageId(state, action: PayloadAction<number>) {
            state.pagesId.images = action.payload;
        },
        setReviewsPageId(state, action: PayloadAction<number>) {
            state.pagesId.reviews = action.payload;
        },
        setSearchMovies(state, action: PayloadAction<IMovie[]>) {
            state.searchMovies = action.payload;
        },
        setFilmRating(state, action) {
            state.filmRating = {};
            state.filmRating = {...action.payload};
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(loadMoviesBySearch.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadMoviesBySearch.rejected, (state) => {
                state.isError = true;
                state.isLoading = false;
            })
            .addCase(loadMoviesBySearch.fulfilled, (state, action: PayloadAction<IMovie[]>) => {
                state.searchMovies = action.payload;
                state.isLoading = false;
                state.isError = false;
            })
            .addCase(loadMovieSequelsAndPrequels.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadMovieSequelsAndPrequels.rejected, (state) => {
                state.isError = true;
                state.isLoading = false;
            })
            .addCase(loadMovieSequelsAndPrequels.fulfilled, (state, action: PayloadAction<IMovie[]>) => {
                state.sequelsAndPrequels = action.payload;
                state.isLoading = false;
                state.isError = false;
            });
    },
});

export const {
    setActiveTabName,
    setTop250PageId,
    setTop100PageId,
    setTopAwaitPageId,
    setImagesPageId,
    setReviewsPageId,
    setSearchMovies,
    setFilmRating,
} = filmsSlice.actions;

export default filmsSlice.reducer;
