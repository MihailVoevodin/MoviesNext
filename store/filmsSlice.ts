import {createSlice, PayloadAction} from '@reduxjs/toolkit';

/**
 * @param pageId - Номер страницы топа фильмов.
 * @param imagesPageId - Номер страницы изображений.
 * @param reviewsPageId - Номер страницы рецензий.
 */
export interface IFilmsState {
    pageId: number;
    imagesPageId: number;
    reviewsPageId: number;
}

const initialState: IFilmsState = {
    pageId: 1,
    imagesPageId: 1,
    reviewsPageId: 1,
};

/**
 * Срез списка фильмов.
 */
const filmsSlice = createSlice({
    name: 'films',
    initialState: initialState,
    reducers: {
        setPageId(state, action: PayloadAction<number>) {
            state.pageId = action.payload;
        },
        setImagesPageId(state, action: PayloadAction<number>) {
            state.imagesPageId = action.payload;
        },
        setReviewsPageId(state, action: PayloadAction<number>) {
            state.reviewsPageId = action.payload;
        },
    },
});

export const {setPageId, setImagesPageId, setReviewsPageId} = filmsSlice.actions;

export default filmsSlice.reducer;
