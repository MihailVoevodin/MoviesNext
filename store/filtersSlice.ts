import {createSlice, PayloadAction, createAsyncThunk} from '@reduxjs/toolkit';
import {IMovie} from 'Common/Models';
import {Services} from 'Common/Services';

export interface IFiltersState {
    movies?: IMovie[];
    genreId: string;
    countryId: string;
    orderId: string;
    typeId: string;
    ratingFrom: string;
    ratingTo: string;
    yearFrom: string;
    yearTo: string;
    keyword: string;
    findMoviesPageId: number;
}

const initialState: IFiltersState = {
    movies: [],
    genreId: '1',
    countryId: '1',
    orderId: 'RATING',
    typeId: 'ALL',
    ratingFrom: '0',
    ratingTo: '10',
    yearFrom: '1000',
    yearTo: '3000',
    keyword: '',
    findMoviesPageId: 1,
};

export const loadFilmsByFilters = createAsyncThunk(
    'filters/getFilmsByFilters',
    async (
        {orderId, genreId, countryId, typeId, ratingFrom, ratingTo, yearFrom, yearTo, keyword, findMoviesPageId}: IFiltersState,
        {rejectWithValue}
    ) => {
        console.log(findMoviesPageId);
        const response = await Services.getFilmsByFilters(
            orderId,
            genreId,
            countryId,
            typeId,
            ratingFrom,
            ratingTo,
            yearFrom,
            yearTo,
            keyword,
            findMoviesPageId
        );

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data.items;
    }
);

/**
 * Срез фильтров поиска фильмов.
 */
const filtersSlice = createSlice({
    name: 'filters',
    initialState: initialState,
    reducers: {
        setGenreId(state, action: PayloadAction<string>) {
            state.genreId = action.payload;
        },
        setCountryId(state, action: PayloadAction<string>) {
            state.countryId = action.payload;
        },
        setOrderId(state, action: PayloadAction<string>) {
            state.orderId = action.payload;
        },
        setTypeId(state, action: PayloadAction<string>) {
            state.typeId = action.payload;
        },
        setRatingFrom(state, action: PayloadAction<string>) {
            state.ratingFrom = action.payload;
        },
        setRatingTo(state, action: PayloadAction<string>) {
            state.ratingTo = action.payload;
        },
        setYearFrom(state, action: PayloadAction<string>) {
            state.yearFrom = action.payload;
        },
        setYearTo(state, action: PayloadAction<string>) {
            state.yearTo = action.payload;
        },
        setKeyword(state, action: PayloadAction<string>) {
            state.keyword = action.payload;
        },
        setFindMoviesPageId(state, action: PayloadAction<number>) {
            state.findMoviesPageId = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder.addCase(loadFilmsByFilters.fulfilled, (state, action) => {
            console.log(action.payload);
            state.movies = action.payload;
        });
    },
});

export const {
    setGenreId,
    setCountryId,
    setOrderId,
    setTypeId,
    setRatingFrom,
    setRatingTo,
    setYearFrom,
    setYearTo,
    setKeyword,
    setFindMoviesPageId,
} = filtersSlice.actions;

export default filtersSlice.reducer;
