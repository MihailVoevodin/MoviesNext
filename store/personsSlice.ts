import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import { IPerson } from "Common/Models";

/**
 * @param persons - Массив личностей по поиску.
 * @param name - Имя введенное в поиск.
 */
export interface IPersonsState {
  persons: IPerson[];
  name: string;
}

const initialState: IPersonsState = {
  persons: [],
  name: ''
};

/**
 * Срез списка личностей.
 */
const personsSlice = createSlice({
  name: 'persons',
  initialState: initialState,
  reducers: {
    setPersonsList(state, action: PayloadAction<IPerson[]>) {
      state.persons = action.payload;
    },
    setName(state, action: PayloadAction<string>) {
      state.name = action.payload;
    },
  },
});

export const {setPersonsList, setName} = personsSlice.actions;

export default personsSlice.reducer;
