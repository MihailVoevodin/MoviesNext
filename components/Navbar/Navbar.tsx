import styles from 'components/Navbar/Navbar.module.scss';
import Link from 'next/link';
import {FC} from 'react';
import {useAppSelector} from 'store/hooks';

/**
 * Компонент навигационной панели.
 */
const Navbar: FC = () => {
    const {pageId} = useAppSelector((state) => state.films);

    const NavigationItems = [
        {id: 0, title: 'Главная', path: '/'},
        {id: 1, title: 'Фильмы', path: `/movies/page/${pageId}`},
        {id: 2, title: 'Личности', path: '/persons'},
    ];

    return (
        <nav>
            <ul className={styles.navbarList}>
                {NavigationItems.map(({id, title, path}) => (
                    <Link key={id} href={path}>
                        {title}
                    </Link>
                ))}
            </ul>
        </nav>
    );
};

export default Navbar;
