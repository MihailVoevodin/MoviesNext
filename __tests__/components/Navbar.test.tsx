import Navbar from 'components/Navbar/Navbar';
import {useRouter} from 'next/router';
import {Provider} from 'react-redux';
import filmsSlice from 'store/films/filmsSlice';
import filtersSlice from 'store/filters/filtersSlice';
import personsSlice from 'store/persons/personsSlice';
import {configureStore} from '@reduxjs/toolkit';
import {fireEvent, render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

jest.mock('next/router', () => ({
    useRouter() {
        return {
            asPath: '/',
        };
    },
}));

jest.spyOn(require('next/router'), 'useRouter');

describe('Navbar display', () => {
    const store = configureStore({
        reducer: {
            films: filmsSlice,
            persons: personsSlice,
            filters: filtersSlice,
        },
    });

    const navbar = () =>
        render(
            <Provider store={store}>
                <Navbar />
            </Provider>
        );

    beforeEach(() => {
        navbar();
    });

    (useRouter as jest.Mock).mockImplementation(() => ({
        useRouter() {
            return {
                asPath: '/',
            };
        },
    }));

    test('correctly render', () => {
        const list = screen.getByRole('list');
        const nav = screen.getByRole('navigation');

        expect(list).toBeInTheDocument();
        expect(nav).toBeInTheDocument();
    });

    test('input typing', () => {
        const input = screen.getByTestId('searchInput');

        expect(input).toBeInTheDocument();
        userEvent.type(input, 'Матрица');
        expect(input).toHaveDisplayValue('Матрица');

        const closeIcon = screen.getByTestId('closeIcon');

        expect(closeIcon).toBeInTheDocument();
        fireEvent.click(closeIcon);
        expect(input).toHaveDisplayValue('');
    });

    test('should match snapshot', () => {
        expect(navbar).toMatchSnapshot();
    });
});
