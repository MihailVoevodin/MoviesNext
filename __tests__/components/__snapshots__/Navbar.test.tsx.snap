// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar display should match snapshot 1`] = `[Function]`;
